### ISSUE TITLE
Refactor the json-ld parser to not use a whole N3 library

### ISSUE THREAD

**Issue by timbl** [linkeddata/rdflib.js#349]
Currently the way JSON-LD parser has been ported into rdflib is a horrible kludge.  It parses it, puts it into a store in a separate N3.js library which is imports, then re-serializes it into something which rdflib.js could parse anyway, and then parse that into the rdflib store.

This involves a huge amount of code and a huge amount of time.

Solution is to change the code of the jsonld parser to write directly to  the rdflib store.

(This may be a dup of an earlier issue but I could not find that)

**Comment by jamsden** [comment]
Should include the serializer too.

**Comment by elf-pavlik** [comment]
Maybe parser could just implement [RDF/JS Sink](http://rdf.js.org/stream-spec/#sink-interface). Big part of RDF/JS goals relates to ease of using various implementations of parsers with various implementations of stores. Actually I see existing JSON-LD parser implementing RDF/JS Sink listed on http://rdf.js.org/

**Comment by josephguillaume** [comment]
Looks like this change has been merged already in Oct 2019 https://github.com/linkeddata/rdflib.js/commit/16daad7d68d805aa29e8e01bb4a16d77b28b6afe, with some issues still remaining #408 #435 #430 

I'm not sure whether the serialiser has been changed or needs to be, and it looks like an RDF/JS Sink approach has not been used, but it seems like this issue can be closed?

**Comment by josephguillaume** [comment]
It appears the serialisation works by converting to ntriples (https://github.com/linkeddata/rdflib.js/blob/master/src/serialize.ts#L72), then to quads, and finally using `jsonld.fromRDF` (https://github.com/linkeddata/rdflib.js/blob/master/src/convert.js#L25), but given the title here mentions the parser, perhaps optimising that process should be a new issue?

**Comment by melvincarvalho** [comment]
I noticed this issue is still open and did some investigation into why N3.js is still bundled.

## Current State

The JSON-LD fix from 2019 (@fletcher91's [16daad7d](https://github.com/linkeddata/rdflib.js/commit/16daad7d)) successfully removed N3.js from that parsing path. However, the dependency remained because it's still used for **N-Quads parsing** in `src/parse.ts:62`.

Additionally, there's a dead import in `src/lists.ts:8` (accidentally copied from parse.ts in 2022) that was never used.

## Possible Solution

Since N-Quads is much simpler than full Turtle/N3 (`<subject> <predicate> <object> [<graph>] .`), I've prototyped a lightweight custom parser (~200 lines) that could replace the N3.js dependency entirely.

**Testing results:**
- Bundle size: 263 KiB → 209 KiB (54 KiB / 20.5% reduction)
- All 293 existing tests passing
- Handles all N-Quads features (literals, language tags, datatypes, graphs, escapes)

## Question

Would a PR to fully remove the N3.js dependency be welcome? Or are there reasons to keep it that I might be missing?

Happy to provide more details or adjust the approach based on feedback.

**Comment by elf-pavlik** [comment]
Since N3.js can parse Turtle, Trig, N-Triples, N-Quads and N3, are you comparing the bundle size with equivalent parser (set of parser) or you just compare it to N-Quads only parser?
If the bundle size is of concern I think comparison would need to happen between the full set of bundled parsing capabilities.
If anyone tried I would also be interested to know how it compares to https://github.com/blake-regalia/graphy.js

I recall around work on Solid Catalog it came up that rdflib.js could fully parse Turtle1.1, I think it was this issue
* https://github.com/linkeddata/rdflib.js/issues/626

If N3.js is already bundled full support of parsing Turtle 1.1 shouldn't be an issue. 

**EDIT**: If I only need to parse JSON-LD I use https://github.com/rubensworks/jsonld-streaming-parser.js and jsonld.js only if i need all the JSON-LD API and Framing.

**Comment by jeswr** [comment]
Closing: the kludge this issue describes was fixed back in 2019 by 16daad7d — the JSON-LD parser now writes directly to the store via the `jsonld` package, with no N3 serialization round-trip (src/jsonldparser.js).

On the follow-up question of why N3.js is still a dependency (thanks @melvincarvalho for the investigation): it currently backs N-Quads parsing (src/parse.ts:62-64) and, with #831, becomes the Turtle/N3 parser for the whole family — so N3.js is staying as an intentional core dependency, and I'm declining the custom N-Quads parser offer: once #831 lands, a second hand-rolled parser would add maintenance surface without reducing the bundle. The stray `n3` import at src/lists.ts:8 will be removed as part of that cleanup.
